Src/Support/Google.Apis.Auth/SignedTokenVerification.cs

```diff
@@ -0,0 +1,249 @@
+using Google.Apis.Util;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using static Google.Apis.Auth.JsonWebSignature;
+
+namespace Google.Apis.Auth
+{
+    internal static class SignedTokenVerification
+    {
+        // See http://oid-info.com/get/2.16.840.1.101.3.4.2.1
+        private const string Sha256Oid = "2.16.840.1.101.3.4.2.1";
+
+        private static readonly DateTime UnixEpoch = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
+        private static readonly CertificateCache s_certificateCache = new CertificateCache();
+
+        internal async static Task VerifySignedTokenAsync<TJswHeader, TJswPayload>(
+            SignedToken<TJswHeader, TJswPayload> signedToken, SignedTokenVerificationOptions options, CancellationToken cancellationToken)
+            where TJswHeader : Header
+            where TJswPayload : Payload
+        {
+            signedToken.ThrowIfNull(nameof(signedToken));
+            options = options == null ? SignedTokenVerificationOptions.Default : new SignedTokenVerificationOptions(options);
+
+            // Start the signature validation task...
+            Task signatureVerificationTask = signedToken.Header.Algorithm switch
+            {
+                "RS256" => VerifyRS256TokenAsync(signedToken, options, cancellationToken),
+                "ES256" => VerifyES256TokenAsync(signedToken, options, cancellationToken),
+                _ => throw new InvalidJwtException("Signing algorithm must be either RS256 or ES256.")
+            };
+
+            // ... let's validate everything else while we wait for signature validation...
+
+            // The signed token issuer should be one of the trusted issuers.
+            if (options.TrustedIssuers.Count > 0 && !options.TrustedIssuers.Contains(signedToken.Payload.Issuer))
+            {
+                var validList = string.Join(", ", options.TrustedIssuers.Select(x => $"'{x}'"));
+                throw new InvalidJwtException($"JWT issuer incorrect. Must be one of: {validList}");
+            }
+            // All audiences on the signed token should be trusted audiences.
+            if (options.TrustedAudiences.Count > 0 && signedToken.Payload.AudienceAsList.Except(options.TrustedAudiences).Any())
+            {
+                throw new InvalidJwtException("JWT contains untrusted 'aud' claim.");
+            }
+            // Issued at and expiration are present.
+            if (signedToken.Payload.IssuedAtTimeSeconds == null || signedToken.Payload.ExpirationTimeSeconds == null)
+            {
+                throw new InvalidJwtException("JWT must contain 'iat' and 'exp' claims");
+            }
+            // Check that the token was issued in the past.
+            var nowSeconds = (options.Clock.UtcNow - UnixEpoch).TotalSeconds;
```

Comment:
I suspect this could be refactored somewhat for readability, but we can do that later. Remind me to have a look once this is in.